pitop/robotics/tilt_roll_head_controller.py

```diff
@@ -0,0 +1,167 @@
+import configparser
+from os.path import (
+    exists,
+    isdir,
+    join,
+)
+from pathlib import Path
+
+from pitop.core.exceptions import UninitializedComponent
+from pitop.core.mixins import (
+    Stateful,
+    Recreatable,
+)
+from pitop.pma import ServoMotor
+from simple_pid import PID
+
+
+class TiltRollHeadController(Stateful, Recreatable):
+    CALIBRATION_FILE_DIR = ".config/pi-top/sdk"
+    CALIBRATION_FILE_NAME = "prax.conf"
+    _initialized = False
+    _roll_servo = None
+    _tilt_servo = None
+
+    def __init__(self, servo_roll_port="S0", servo_tilt_port="S3", name="tilt_roll"):
+        self.name = name
+        self._roll_servo = ServoMotor(servo_roll_port)
+        self._tilt_servo = ServoMotor(servo_tilt_port)
+        self.__calibration_file_path = join(str(Path.home()), self.CALIBRATION_FILE_DIR, self.CALIBRATION_FILE_NAME)
+        self._initialized = True
+
+        self._head_roll_pid = PID(Kp=3.0,
+                                  Ki=1.5,
+                                  Kd=0.25,
+                                  setpoint=0,
+                                  output_limits=(-100, 100)
+                                  )
+
+        Stateful.__init__(self, children=['_roll_servo', '_tilt_servo'])
+        Recreatable.__init__(self, config_dict={'servo_roll_port': servo_roll_port, 'servo_tilt_port': servo_tilt_port, 'name': name})
+
+    def is_initialized(fcn):
+        def check_initialization(self, *args, **kwargs):
+            if not self._initialized:
+                raise UninitializedComponent("TiltRollHeadController not initialized")
+            return fcn(self, *args, **kwargs)
+        return check_initialization
+
+    @property
+    @is_initialized
+    def roll_servo(self):
+        return self._roll_servo
+
+    @property
+    @is_initialized
+    def tilt_servo(self):
+        return self._tilt_servo
+
+    def stop(self):
+        self.tilt_servo.sweep(speed=0)
+        self.roll_servo.sweep(speed=0)
+
+    def track_head_angle(self, angle, flipped=True):
+        if flipped:
+            angle = -angle
+        current_angle = self.roll_servo.current_angle
+        state = current_angle - angle
+        if abs(state) < 1.0:
+            self.roll_servo.sweep(speed=0)
+        else:
+            servo_speed = self._head_roll_pid(state)
+            self.roll_servo.sweep(speed=servo_speed)
+
+    @is_initialized
+    def calibrate(self, save=True, reset=False):
```

Comment:
need to delete calibration from here, too much duplication with pantiltcontroller